File: reference-implementations/py/src/cardano/cardano.py
# -*- encoding: utf-8 -*-
"""
Cardano Anoncreds Method

"""

from blockfrost import BlockFrostApi, ApiError, ApiUrls
from pycardano import * 
from textwrap import wrap
from threading import Timer
from  pprint import pp
import os
import json
import random
from models.schema import Schema
from models.object_metadata import ObjectMetadata
from models.cred_def import CredDef
from models.rev_reg import RevReg
from models.rev_reg_entry import RevRegEntry
import hashlib
from datetime import datetime, timezone
from typing import Optional

NETWORK = Network.TESTNET
MINIMUN_BALANCE = 5000000
TRANSACTION_AMOUNT = 1000000
MINIMUN_UTXO = 20
ALLOWS_ASYNC = True
QUEUE_DURATION = 60

class Cardano:
    """
    Environment variables required:
        - BLOCKFROST_API_KEY = API KEY from Blockfrots  https://blockfrost.io
        - CARDANO_ADDRESS_CBORHEX = Private Key of address as CBOR Hex. Must be an Enterprice address (no Staking part) as PaymentSigningKeyShelley_ed25519
    
        - Support queued operations. No Tx_hash can be retrieved syncronously.
        - To guarantee synchronous operation, enouth UTXOs must be available in the address to process simultaneous operations.
        - If not, the operation will be queued and processed when UTXOs are available. Need to define how to retrieve tx_hash in this case.
        - spreadUTxOs() will spread UTXOs in the address to avoid this situation.

        TODO:
        - validate anoncreds object against models
        - validate signature
        - low balance warning

    """

    def __init__(self):
        self.pending_tx = []
        self.available_utxos = []
        self.timer = Timer(QUEUE_DURATION, self.flushQueue)
        self.ledger_cache = {}

        try:
            blockfrost_project_id=os.environ['BLOCKFROST_API_KEY']
            self.api = BlockFrostApi(
            project_id=blockfrost_project_id,
            base_url=ApiUrls.preview.value
            )
            self.context = BlockFrostChainContext(blockfrost_project_id,NETWORK, ApiUrls.preview.value)
        except KeyError:
             print("Missing environment variable BLOCKFROST_API_KEY")
             exit(1)
        try:
            self.payment_signing_key = PaymentSigningKey.from_cbor(os.environ['CARDANO_ADDRESS_CBORHEX'])
            payment_verification_key = PaymentVerificationKey.from_signing_key(self.payment_signing_key)
            self.payment_addr = Address(payment_verification_key.hash(), None, network=NETWORK)
        except KeyError:
            print("Missing environment variable CARDANO_ADDRESS_CBORHEX")
            exit(1)

        balance = self.getaddressBalance()
        if balance < MINIMUN_BALANCE:
            print("Insuficient funds in address",self.payment_addr.encode())
            exit(1)
        else:
            print("Address",self.payment_addr.encode(),"has",balance,"ADA")
            self.spreadUTxOs()
            
    def registerSchema(self,schema: Schema, publisher_DID: str, signature: str):
        
        if self.validateSignature:
            object_metadata: ObjectMetadata = {
                "resourceURI": "",
                "resourceName": schema["name"],
                "resourceFamily": "anoncreds",
                "resourceType": "SCHEMA",
                "resourceVersion": "v1",
                "mediaType": "application/json",
                "created": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "checkSum": hashlib.md5(json.dumps(schema).encode('utf-8')).hexdigest(),
                "publisherId": publisher_DID,
                "publisherSignature": signature
            }
            tx_id = self.publishAnoncredObject(schema,object_metadata)

            return publisher_DID + "/resources/" + str(tx_id)
        else: return "Invalid Signature"

    def registerCredDef(self,cred_def: CredDef, publisher_DID: str, signature: str):
        if self.validateSignature:
            object_metadata: ObjectMetadata = {
                "resourceURI": "",
                "resourceName": "CL",
                "resourceFamily": "anoncreds",
                "resourceType": "CRED_DEF",
                "resourceVersion": "v1",
                "mediaType": "application/json",
                "created": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "checkSum": hashlib.md5(json.dumps(cred_def).encode('utf-8')).hexdigest(),
                "publisherId": publisher_DID,
                "publisherSignature": signature
            }
            tx_id = self.publishAnoncredObject(cred_def,object_metadata)

            return publisher_DID + "/resources/" + str(tx_id)
        else: return "Invalid Signature"
    
    def registerRevReg(self,rev_reg: RevReg, publisher_DID: str, signature: str):
        if self.validateSignature:
            object_metadata: ObjectMetadata = {
                "resourceURI": "",
                "resourceName": "CL_ACUMM",
                "resourceFamily": "anoncreds",
                "resourceType": "REV_REG",
                "resourceVersion": "v1",
                "mediaType": "application/json",
                "created": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "checkSum": hashlib.md5(json.dumps(rev_reg).encode('utf-8')).hexdigest(),
                "publisherId": publisher_DID,
                "publisherSignature": signature
            }
            tx_id = self.publishAnoncredObject(rev_reg,object_metadata)

            return publisher_DID + "/resources/" + str(tx_id)
        else: return "Invalid Signature"
    
    def registerRevRegEntry(self, rev_reg_entry: RevRegEntry, publisher_DID: str, signature: str):
        if self.validateSignature:
            object_metadata: ObjectMetadata = {
                "resourceURI": "",
                "resourceName": "CL_ACUMM",
                "resourceFamily": "anoncreds",
                "resourceType": "REV_REG_ENTRY",
                "resourceVersion": "v1",
                "mediaType": "application/json",
                "created": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
                "checkSum": hashlib.md5(json.dumps(rev_reg_entry).encode('utf-8')).hexdigest(),
                "publisherId": publisher_DID,
                "publisherSignature": signature
            }
            rev_reg_id = rev_reg_entry["revocRegDefId"]
            object_id = rev_reg_id.split("/")[-1]
            meta = self.getObject(object_id)
            prefix = int(meta[0]['label'])

            tx_id = self.publishAnoncredObject(rev_reg_entry,object_metadata, prefix)

            return publisher_DID + "/resources/" + str(tx_id)
        else: return "Invalid Signature"

    def resolveObject(self, resource_URI):
        object_id = resource_URI.split("/")[-1]
        meta = self.getObject(object_id)

        meta_json = json.loads(''.join(meta[0]['json_metadata']))

        meta_json["ResourceObjectMetadata"]["resourceURI"] = resource_URI

        # For RevReg also loads all RevRegEntries
        if meta_json["ResourceObjectMetadata"]["resourceType"] == "REV_REG":
            meta_key = meta[0]["label"]
            entries = self.getRevocationEntries(meta_key, resource_URI, meta_json["ResourceObjectMetadata"]["publisherId"])
            meta_json["RevRegEntries"] = entries

        return meta_json

    def queryObject(self, resource_URI):
        return
    
    def publishAnoncredObject(self, anoncred_object,anoncred_object_metadata, metadata_prefix: Optional[int] = None) -> str:
        if not metadata_prefix:  metadata_prefix = random.randint(0, 2**32-1)
        object = {
                "ResourceObject": anoncred_object,
                "ResourceObjectMetadata": anoncred_object_metadata
            }
        meta = {
            metadata_prefix: wrap(json.dumps(object), 64)
        }
        return self.processTransaction(meta)
    
    def processTransaction(self, meta):
        
        if not self.timer.is_alive():
            self.getUTXOs()
            self.pending_tx.append(meta)
            tx_hash = self.submitTransaction(meta)
            self.timer = Timer(90, self.flushQueue)
            self.timer.start()
        elif len(self.available_utxos) > 0:
            self.pending_tx.append(meta)
            tx_hash = self.submitTransaction(meta)
        else:
            tx_hash = "Queued operation"
            if ALLOWS_ASYNC: 
                self.pending_tx.append(meta)
        return tx_hash

    def submitTransaction(self,meta):
        # Build transaction
        try:
            builder = TransactionBuilder(self.context)
            # select utxos
            utxo_sum = 0
            utxo_to_remove = []
            # print("Qty of utxo",len(self.available_utxos))
            for u in self.available_utxos:
                utxo_sum = utxo_sum + int(u.amount[0].quantity)
                builder.add_input(
                    UTxO(
                        TransactionInput.from_primitive([u.tx_hash, u.tx_index]),
                        TransactionOutput(address=Address.from_primitive(u.address), amount=int(u.amount[0].quantity))
                    )
                )
                utxo_to_remove.append(u)
                if utxo_sum > (TRANSACTION_AMOUNT + 2000000): break
            for ur in utxo_to_remove: self.available_utxos.remove(ur)
            builder.add_output(TransactionOutput(self.payment_addr,Value.from_primitive([TRANSACTION_AMOUNT])))
            builder.auxiliary_data = AuxiliaryData(Metadata(meta))
            signed_tx = builder.build_and_sign([self.payment_signing_key], change_address=self.payment_addr)
            # Submit transaction
            self.context.submit_tx(signed_tx.to_cbor())
            self.ledger_cache[str(signed_tx.id)] = [
                {
                    "label": list(meta.keys())[0],
                    "json_metadata": meta[list(meta.keys())[0]]
                }
            ]
            self.pending_tx.remove(meta)
            return signed_tx.id
        except Exception as e:
            print(e)
            self.available_utxos = []

    def flushQueue(self):
        print("Flushing Queue")
        self.getUTXOs()
        for m in self.pending_tx.copy():
            self.submitTransaction(m)

    def getObject(self, tx_id):
        if tx_id in self.ledger_cache:
            meta = self.ledger_cache[tx_id]
        else:
            meta = self.api.transaction_metadata(tx_id, return_type='json')
        return meta

    def getRevocationEntries(self, meta_key, rev_reg_id, publisher_DID):
        metas = self.api.metadata_label_json(meta_key, return_type='json')
        accumulators = []
        
        for m in metas:
            meta_json = json.loads(''.join(list(m["json_metadata"].values())))
            if meta_json["ResourceObjectMetadata"]["resourceType"] == "REV_REG_ENTRY" and meta_json["ResourceObject"]["revocRegDefId"] == rev_reg_id and  meta_json["ResourceObjectMetadata"]["publisherId"] == publisher_DID:
                # TODO VERIFY SIGNATURE
                accumulators.append(meta_json["ResourceObject"])
        return accumulators

    def getaddressBalance(self):
        try:
            address = self.api.address(address=self.payment_addr.encode())
            return int(address.amount[0].quantity)
        except ApiError as e:
            return 0

    def getUTXOs(self):
        try:
            self.available_utxos = self.api.address_utxos(self.payment_addr.encode())
        except ApiError as e:
            print("error",e)

    def spreadUTxOs(self):

        utxos = self.api.address_utxos(self.payment_addr.encode())
        balance = self.getaddressBalance()
        if balance > MINIMUN_BALANCE and len(utxos) < MINIMUN_UTXO:
            print("Spreading UTXOs")
            tx_amount = int((balance - 10000000) / MINIMUN_UTXO)
            print(tx_amount)
            builder = TransactionBuilder(self.context)
            builder.add_input_address(self.payment_addr)
            for _ in range(MINIMUN_UTXO):
                builder.add_output(TransactionOutput(self.payment_addr,Value.from_primitive([tx_amount])))
            signed_tx = builder.build_and_sign([self.payment_signing_key], change_address=self.payment_addr)
            # Submit transaction
            self.context.submit_tx(signed_tx.to_cbor())
    
    def validateSignature(self, object, publisherDID, signature):
        # TODO
        # resolve DID
        # get public key from DID Doc
        # verify signature
        return True

File: reference-implementations/ts/src/cardano/cardano.ts
/* 
Environment variables required:
    - BLOCKFROST_API_KEY = API KEY from Blockfrots  https://blockfrost.io
    - CARDANO_ADDRESS_CBORHEX = Private Key of address as CBOR Hex. Must be an Enterprice address (no Staking part) as PaymentSigningKeyShelley_ed25519
    - WARNING!!! Storing private keys in environment variables is not secure at all!!!. This is only for testing purposes.

    - Support queued operations. No Tx_hash can be retrieved syncronously.
    - To guarantee synchronous operation, enouth UTXOs must be available in the address to process simultaneous operations.
    - If not, the operation will be queued and processed when UTXOs are available. Need to define how to retrieve tx_hash in this case.
    - spreadUTxOs() will spread UTXOs in the address to avoid this situation.

TODO:
    - metadata label pagination
    - improve error handling
    - validate anoncreds object against models
    - validate signature
    - low balance warning
    - implementy query for objetcs
*/

import { BlockFrostAPI } from '@blockfrost/blockfrost-js'
import { Options } from '@blockfrost/blockfrost-js/lib/types'
import * as cardanoWasm from '@emurgo/cardano-serialization-lib-nodejs'
import {randomBytes } from 'crypto'
import cbor from 'cbor'
import {MD5} from 'crypto-js'
import { ISchema } from './models/ISchema'
import { IObjectMetadata } from './models/IObjectMetadata'
import { ICredDef } from './models/ICredDef'
import { IRevReg } from './models/IRevReg'
import { IRevRegEntry } from './models/IRevRegEntry'
import { IMetadata } from './models/IMetadata'
import { mnemonicToEntropy } from 'bip39';

const NETWORK = "preview"
const MINIMUN_BALANCE = 5000000
const TRANSACTION_AMOUNT = 1000000
const MINIMUM_UTXO = 20
const ALLOWS_ASYNC = true
const QUEUE_DURATION = 90


export default class Cardano {

    private readonly blockfrostAPI: BlockFrostAPI
    private paymentAddress: string
    private ledgerCache: Map<string, any>
    private availableUTXOs: any[]
    private timer: NodeJS.Timeout | null
    private pendingTx: any[]
    private privateKey: cardanoWasm.PrivateKey
    
  
    constructor () 
    {
        const blockfrostProjectId = process.env.BLOCKFROST_API_KEY
        this.ledgerCache = new Map()
        this.availableUTXOs = []
        this.timer = null
        this.pendingTx = []
        this.paymentAddress = ""
        this.blockfrostAPI = new BlockFrostAPI(
            { 
                projectId: blockfrostProjectId,
                network: NETWORK
            } as Options
        )

        if (process.env.CARDANO_ADDRESS_CBORHEX){
            const cardanoPrivateAddress = process.env.CARDANO_ADDRESS_CBORHEX
            this.privateKey = cardanoWasm.PrivateKey.from_normal_bytes(new Uint8Array(cbor.decode(cardanoPrivateAddress!)))
            this.paymentAddress = cardanoWasm.EnterpriseAddress.new(
                cardanoWasm.NetworkInfo.testnet().network_id(),
                cardanoWasm.StakeCredential.from_keyhash(this.privateKey.to_public().hash())
              ).to_address().to_bech32()
            this.getUTXOs()
        } else {
            this.privateKey  = cardanoWasm.PrivateKey.generate_ed25519()
            this.createEnterpriseAddress()
        }
        
        
        

        


        
    }


    public async registerSchema(schema: ISchema, publisher_DID: string, signature: string): Promise<string> {
        if (await this.validateSignature(schema, publisher_DID, signature)){
            const objectMetadata: IObjectMetadata = {
                resourceURI: "",
                resourceName: schema["name"],
                resourceFamily: "anoncreds",
                resourceType: "SCHEMA",
                resourceVersion: "v1",
                mediaType: "application/json",
                created: (new Date()).toISOString(),
                checkSum: MD5(JSON.stringify(schema)).toString(),
                publisherId: publisher_DID,
                publisherSignature: signature
            }
            const txId = await this.publishAnoncredObject(schema,objectMetadata)
            return publisher_DID + "/resources/" + txId
        } else { return "Invalid Signature"}
    }

    public async registerCredDef(credDef: ICredDef, publisher_DID: string, signature: string): Promise<string> {
        if (await this.validateSignature(credDef, publisher_DID, signature)){
            const objectMetadata: IObjectMetadata = {
                resourceURI: "",
                resourceName: "CL",
                resourceFamily: "anoncreds",
                resourceType: "CRED_DEF",
                resourceVersion: "v1",
                mediaType: "application/json",
                created: (new Date()).toISOString(),
                checkSum: MD5(JSON.stringify(credDef)).toString(),
                publisherId: publisher_DID,
                publisherSignature: signature
            }
            const txId = await this.publishAnoncredObject(credDef,objectMetadata)
            return publisher_DID + "/resources/" + txId
        } else { return "Invalid Signature"}
    }

    public async registerRevReg(revReg: IRevReg, publisher_DID: string, signature: string): Promise<string> {
        if (await this.validateSignature(revReg, publisher_DID, signature)){
            const objectMetadata: IObjectMetadata = {
                resourceURI: "",
                resourceName: "CL_ACUMM",
                resourceFamily: "anoncreds",
                resourceType: "REV_REG",
                resourceVersion: "v1",
                mediaType: "application/json",
                created: (new Date()).toISOString(),
                checkSum: MD5(JSON.stringify(revReg)).toString(),
                publisherId: publisher_DID,
                publisherSignature: signature
            }
            const txId = await this.publishAnoncredObject(revReg,objectMetadata)
            return publisher_DID + "/resources/" + txId
        } else { return "Invalid Signature"}
    }

    public async registerRevRegEntry(revRegEntry: IRevRegEntry, publisher_DID: string, signature: string): Promise<string> {
        if (await this.validateSignature(revRegEntry, publisher_DID, signature)){
            const objectMetadata: IObjectMetadata = {
                resourceURI: "",
                resourceName: "CL_ACUMM",
                resourceFamily: "anoncreds",
                resourceType: "REV_REG_ENTRY",
                resourceVersion: "v1",
                mediaType: "application/json",
                created: (new Date()).toISOString(),
                checkSum: MD5(JSON.stringify(revRegEntry)).toString(),
                publisherId: publisher_DID,
                publisherSignature: signature
            }
            const revRegId = revRegEntry.revocRegDefId
            const objectId = revRegId.split("/").slice(-1)[0]
            const meta = await this.getObject(objectId)
            const prefix = parseInt(meta[0]['label'])

            const txId = await this.publishAnoncredObject(revRegEntry,objectMetadata, prefix)
            return publisher_DID + "/resources/" + txId
        } else { return "Invalid Signature"}
    }

    public async resolveObject(resourceURI: string): Promise<IMetadata> {
        const objectId = resourceURI.split("/").slice(-1)[0]
        const meta = await this.getObject(objectId)
        let metaJson = JSON.parse(meta[0].json_metadata.join(''))
        metaJson["ResourceObjectMetadata"]["resourceURI"] = resourceURI

        // For RevReg also loads all RevRegEntries
        if (metaJson.ResourceObjectMetadata.resourceType === "REV_REG") {
            const metaKey = meta[0].label
            const entries = await this.getRevocationEntries(metaKey, resourceURI, metaJson.ResourceObjectMetadata.publisherId)
            metaJson["RevRegEntries"] = entries
        }

        
        return metaJson
    }


    async publishAnoncredObject(anoncredObject: ISchema | ICredDef | IRevReg | IRevRegEntry,anoncredObjectMetadata: IObjectMetadata, metadataPrefix?: number): Promise<string>{
        if (!metadataPrefix) {
            metadataPrefix = randomBytes(4).readUInt32BE(0)
        }
        const object = {
            ResourceObject: anoncredObject,
            ResourceObjectMetadata: anoncredObjectMetadata
        }
        // let meta: Map<number, Array<string>> = new Map()
        let meta: any = {}
        meta[metadataPrefix] = JSON.stringify(object).match(/(.{1,64})/g)
        return await this.processTransaction(meta)

    }


    async processTransaction(meta: any): Promise<string> {
        let txHash = ""
        if (this.timer === null) {
            this.getUTXOs()
            this.pendingTx.push(meta)
            txHash = await this.submitTransaction(meta)
            this.timer = setTimeout(async () => await this.flushQueue(), QUEUE_DURATION * 1000)
        } else if (this.availableUTXOs.length > 0) {
            this.pendingTx.push(meta)
            txHash = await this.submitTransaction(meta)
        } else {
            txHash = "queued_operation"
            if (ALLOWS_ASYNC === true) {
                console.log("Queued operation")
                this.pendingTx.push(meta)
            }
        }
        return txHash
    }

    async submitTransaction(meta: any): Promise<string> {

        try {
            const latestEpoch = await this.blockfrostAPI.epochsLatest()
            const protocolParams = await this.blockfrostAPI.epochsParameters(latestEpoch.epoch)
            const latestBlock = await this.blockfrostAPI.blocksLatest()
            const linearFee = cardanoWasm.LinearFee.new(
                cardanoWasm.BigNum.from_str(protocolParams.min_fee_a.toString()),
                cardanoWasm.BigNum.from_str(protocolParams.min_fee_b.toString())
            )
            const txBuilderCfg = cardanoWasm.TransactionBuilderConfigBuilder.new()
                .fee_algo(linearFee)
                .pool_deposit(cardanoWasm.BigNum.from_str(protocolParams.pool_deposit))
                .key_deposit(cardanoWasm.BigNum.from_str(protocolParams.key_deposit))
                .max_value_size(Number(protocolParams.max_val_size))
                .max_tx_size(protocolParams.max_tx_size)
                .coins_per_utxo_word(cardanoWasm.BigNum.from_str(protocolParams.coins_per_utxo_size!))
                .build()

            const txBuilder = cardanoWasm.TransactionBuilder.new(txBuilderCfg)
            // Select UTxOs
            let utxoSum = 0
            let utxoToRemove: any[] = []
            for (const utxo of this.availableUTXOs) {
                utxoSum = utxoSum + parseInt(utxo.amount[0].quantity)
                txBuilder.add_input(
                cardanoWasm.Address.from_bech32(this.paymentAddress),
                cardanoWasm.TransactionInput.new(
                    cardanoWasm.TransactionHash.from_bytes(
                    Buffer.from(utxo.tx_hash, 'hex')),
                    utxo.tx_index
                ),
                cardanoWasm.Value.new(cardanoWasm.BigNum.from_str(utxo.amount[0].quantity))
                );
                utxoToRemove.push(utxo)
                if (utxoSum > (TRANSACTION_AMOUNT + 2000000)){ break}
            }
            
            txBuilder.add_output(
                cardanoWasm.TransactionOutput.new(
                    cardanoWasm.Address.from_bech32(this.paymentAddress),
                cardanoWasm.Value.new(cardanoWasm.BigNum.from_str(TRANSACTION_AMOUNT.toString()))    
                ),
            );
            

            const auxData = cardanoWasm.AuxiliaryData.new()
            const metadata = cardanoWasm.encode_json_str_to_metadatum(JSON.stringify(meta[Object.keys(meta)[0]]),cardanoWasm.MetadataJsonSchema.NoConversions)
            const transactionMetadata = cardanoWasm.GeneralTransactionMetadata.new()
            transactionMetadata.insert(cardanoWasm.BigNum.from_str(Object.keys(meta)[0].toString()), metadata)
            auxData.set_metadata(transactionMetadata)
            txBuilder.set_auxiliary_data(auxData)
            txBuilder.set_ttl(latestBlock.slot! + 600)

            txBuilder.add_change_if_needed(cardanoWasm.Address.from_bech32(this.paymentAddress))
            const txBody = txBuilder.build();
            const txHash = cardanoWasm.hash_transaction(txBody);


            const witnesses = cardanoWasm.TransactionWitnessSet.new();
            const vkeyWitnesses = cardanoWasm.Vkeywitnesses.new();
            const vkeyWitness = cardanoWasm.make_vkey_witness(txHash, this.privateKey);
            vkeyWitnesses.add(vkeyWitness);
            witnesses.set_vkeys(vkeyWitnesses);
            // serialize CBOR transaction
            const transaction = cardanoWasm.Transaction.new(
                txBody,
                witnesses,
                auxData
            );
            this.blockfrostAPI.txSubmit(transaction.to_bytes())
            this.ledgerCache.set(this.toHexString(txHash.to_bytes()), [
                {
                    label: Object.keys(meta)[0],
                    json_metadata: meta[Object.keys(meta)[0]]
                }
            ])
            this.pendingTx.splice(this.pendingTx.indexOf(meta), 1)
            utxoToRemove.forEach(utxo => this.availableUTXOs.splice(this.availableUTXOs.indexOf(utxo), 1))
            return this.toHexString(txHash.to_bytes())
        } catch (error) {
            console.log(error)
            this.availableUTXOs = []
            if (this.timer === null) {
                this.timer = setTimeout(async () => await this.flushQueue(), QUEUE_DURATION * 1000)
            }
            return "tx_queued."
        }

    }

    async flushQueue(): Promise<void> {
        console.log("Flushing Queue")
        await this.getUTXOs()
        this.pendingTx.forEach(async m => {
            await this.submitTransaction(m)
        })
        this.timer = null
            
    }

    async getObject(txID: string): Promise<any>{
        if (this.ledgerCache.has(txID)) {
            return this.ledgerCache.get(txID)
        } else {
            return await this.blockfrostAPI.txsMetadata(txID)
        }
    }

    async getRevocationEntries(metaKey: string, revRegId: string, publisherDID: string): Promise<IRevRegEntry[]>{
        const metas = await this.blockfrostAPI.metadataTxsLabel(metaKey)
        let accumulators: IRevRegEntry[] = []
        metas.forEach(m => {
            let meta = ""

            Object.entries(m.json_metadata!).forEach(([k,v]) => {
                meta += v
            })
            let metaJson = JSON.parse(meta)
              
            if (
                metaJson.ResourceObjectMetadata.resourceType === "REV_REG_ENTRY" && 
                metaJson.ResourceObject.revocRegDefId === revRegId &&
                metaJson.ResourceObjectMetadata.publisherId === publisherDID
            ) {
                    // TODO VERIFY SIGNATURE
                    accumulators.push(metaJson.ResourceObject)
            }
        })
        return accumulators
    }
        

    async getAddressbalance (): Promise<number> {
        try {
            const address = await this.blockfrostAPI.addresses(this.paymentAddress)
            return +address.amount[0].quantity
        } catch (error) {
            return 0
        }
      }

    async getUTXOs(): Promise<void> {
        try {
            this.availableUTXOs = await this.blockfrostAPI.addressesUtxos(this.paymentAddress)
        } catch (error) {
            console.log(error)
        }
    }

    async validateSignature(object: ISchema | ICredDef | IRevReg | IRevRegEntry, publisherDID: string, signature: string): Promise<boolean> {
        // # TODO
        // # resolve DID
        // # get public key from DID Doc
        // # verify signature
        return true
    }

    private toHexString (byteArray: Uint8Array) {
        return Array.from(byteArray, (byte) => {
          return (`0${(byte & 0xFF).toString(16)}`).slice(-2);
        }).join('');
      }

    async createEnterpriseAddress(): Promise<void> {

        this.privateKey  = cardanoWasm.PrivateKey.generate_ed25519()
        this.paymentAddress = cardanoWasm.EnterpriseAddress.new(
            cardanoWasm.NetworkInfo.testnet().network_id(),
            cardanoWasm.StakeCredential.from_keyhash(this.privateKey.to_public().hash())
          ).to_address().to_bech32()
        const cborhex = cbor.encode(this.privateKey.as_bytes().buffer).toString('hex')
        console.log("Enterprise address: " + this.paymentAddress)
        console.log("Private Key CBORHex: " + cborhex)
        let balance = 0 
        while (await this.getAddressbalance() < MINIMUN_BALANCE) {
            balance = await this.getAddressbalance()
            console.log("Waiting for funds...")
            await new Promise(r => setTimeout(r, 30000))
        }
        console.log("Funds received: " + balance/1000, "ADA")
        await this.getUTXOs()
        console.log("Qty of UTXOs: " + this.availableUTXOs.length)
        await this.spreadUTxOs()
        console.log("Waiting for blockchain confirmations...")
        await new Promise(r => setTimeout(r, 60000))
        await this.getUTXOs()
        console.log("Qty of UTXOs: ", this.availableUTXOs.length)

    }    

    async spreadUTxOs(): Promise<void> {
        await this.getUTXOs()
        let balance = await this.getAddressbalance()
        if (balance > MINIMUN_BALANCE && this.availableUTXOs.length < MINIMUM_UTXO){
            console.log("Spreading UTXOs")
            const tx_amount = Math.floor((balance - 10000000) / MINIMUM_UTXO)
            const latestEpoch = await this.blockfrostAPI.epochsLatest()
            const protocolParams = await this.blockfrostAPI.epochsParameters(latestEpoch.epoch)
            const latestBlock = await this.blockfrostAPI.blocksLatest()
            const linearFee = cardanoWasm.LinearFee.new(
                cardanoWasm.BigNum.from_str(protocolParams.min_fee_a.toString()),
                cardanoWasm.BigNum.from_str(protocolParams.min_fee_b.toString())
            )
            const txBuilderCfg = cardanoWasm.TransactionBuilderConfigBuilder.new()
                .fee_algo(linearFee)
                .pool_deposit(cardanoWasm.BigNum.from_str(protocolParams.pool_deposit))
                .key_deposit(cardanoWasm.BigNum.from_str(protocolParams.key_deposit))
                .max_value_size(Number(protocolParams.max_val_size))
                .max_tx_size(protocolParams.max_tx_size)
                .coins_per_utxo_word(cardanoWasm.BigNum.from_str(protocolParams.coins_per_utxo_size!))
                .build()

            const txBuilder = cardanoWasm.TransactionBuilder.new(txBuilderCfg)
            for (const utxo of this.availableUTXOs) {
                txBuilder.add_input(
                cardanoWasm.Address.from_bech32(this.paymentAddress),
                cardanoWasm.TransactionInput.new(
                    cardanoWasm.TransactionHash.from_bytes(
                    Buffer.from(utxo.tx_hash, 'hex')),
                    utxo.tx_index
                ),
                cardanoWasm.Value.new(cardanoWasm.BigNum.from_str(utxo.amount[0].quantity))
                );
            }
            for (const x of Array(MINIMUM_UTXO).keys()) {
                txBuilder.add_output(
                    cardanoWasm.TransactionOutput.new(
                        cardanoWasm.Address.from_bech32(this.paymentAddress),
                    cardanoWasm.Value.new(cardanoWasm.BigNum.from_str(tx_amount.toString()))    
                    ),
                );
            }

            txBuilder.set_ttl(latestBlock.slot! + 600)

            txBuilder.add_change_if_needed(cardanoWasm.Address.from_bech32(this.paymentAddress))
            const txBody = txBuilder.build();
            const txHash = cardanoWasm.hash_transaction(txBody);

            const witnesses = cardanoWasm.TransactionWitnessSet.new();
            const vkeyWitnesses = cardanoWasm.Vkeywitnesses.new();
            const vkeyWitness = cardanoWasm.make_vkey_witness(txHash, this.privateKey);
            vkeyWitnesses.add(vkeyWitness);
            witnesses.set_vkeys(vkeyWitnesses);
            // serialize CBOR transaction
            const transaction = cardanoWasm.Transaction.new(
                txBody,
                witnesses
            );
            this.blockfrostAPI.txSubmit(transaction.to_bytes())
        }
    }
}
